dialogue-core/src/main/java/com/palantir/dialogue/core/QueuedChannel.java

```diff
@@ -78,9 +81,11 @@
     @SuppressWarnings("FutureReturnValueIgnored")
     QueuedChannel(LimitedChannel delegate, int maxQueueSize, DispatcherMetrics metrics) {
         this.delegate = new NeverThrowLimitedChannel(delegate);
-        this.queuedCalls = new LinkedBlockingDeque<>(maxQueueSize);
+        // Do _not_ call size on a ConcurrentLinkedDeque. Unlike other collections, size is an O(n) operation.
+        this.queuedCalls = new ConcurrentLinkedDeque<>();
```

Comment:
I considered subclassing this to throw if `size()` is called. Thoughts?

Longer term we may want to consider something along the lines of https://github.com/undertow-io/undertow/blob/21035b2ee38e703d82dc74292853acfbdbffc46b/core/src/main/java/io/undertow/util/FastConcurrentDirectDeque.java to remove items from the queue immediately (O(1)) when `future.cancel` is invoked. That's a journey for another day though.

Reply:
+1 to subclassing, and if possible could your subclass actually just keep track of the queueSizeEstimate? It would clear out a few incrementAndGet/decrementAndGet calls from this file

Reply:
It could, but as mentioned in the PR description we would lose the ability to pull the head element and attempt to execute it without modifying the estimated size, so the queue would slowly grow when the queue is supposed to be full